README
README
======

Class::DBI::Plugin::DateTime allows you to easily use DateTime with your
Class::DBI objects.


Changes
Changes
=======

0.04 28 Jan 2006
  - Allow explicit specification of implementation to use via
    Class::DBI::Plugin::DateTime qw(Pg);
  - Pg implementation of has_interval() has_duration() wasn't working. Fixed.

0.03 25 Nov 2005
  - Add Class::DBI::Plugin::DateTime to auto-select the implementation class
  - use Module::Pluggable

0.02 17 Nov 2005
  - No code change
  - Minor fix on MySQL.pm docs
  - Properly set abstract on Build.PL

0.01 17 Nov 2005
  - Initial CPAN release


lib/Class/DBI/Plugin/DateTime.pm
# $Id: DateTime.pm 9 2006-01-28 04:54:17Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Class::DBI::Plugin::DateTime;
use strict;
use Module::Pluggable
    sub_name    => 'impls',
    search_path => [ 'Class::DBI::Plugin::DateTime' ],
    except      => [ 'Class::DBI::Plugin::DateTime::Base' ],
    require     => 0,
;
use vars qw($VERSION);
BEGIN {
    $VERSION = '0.04';
}

sub import
{
    my $class = shift;
    my $type  = shift;

    my ($caller) = caller();
    if (!$type) {
        $type  = $caller->db_Main->{Driver}{Name};
    }
    my $impl  = $class->_select_impl($type);

    $impl->_do_export($caller);
}

sub _select_impl
{
    my $class = shift;
    my $type  = lc(shift);

    foreach my $impl ($class->impls) {
        my $p = lc((split(/::/, $impl))[-1]);
        if ($p eq $type) {
            eval "require $impl";
            die if $@;
            return $impl;
        }
    }
}

1;

__END__

=head1 NAME

Class::DBI::Plugin::DateTime - Use DateTime Objects As Columns

=head1 SYNOPSIS

  package MyCDBI;
  # call set_db first
  use Class::DBI::Plugin::DateTime;

  # setup columns, depending on db type

=head1 DESCRIPTION

Class::DBI::Plugin::DateTime is a convenience interface to Class::DBI::Plguin::DateTime::* objects. It auto-detects the connectin type being used, and loads
the appropriate plugin. Note that you need to set up the database connection
before loading this module into your Class::DBI based module.

If you don't want want to bother with this, you can simply use the appropriate
class directly (e.g. Class::DBI::Plugin::DateTime::Pg)

=head1 AUTHOR

Copyright (c) 2005 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>. All rights reserved.

Development funded by Brazil Ltd E<lt>http://b.razil.jpE<gt>

=cut

lib/Class/DBI/Plugin/DateTime/Pg.pm
# $Id: Pg.pm 9 2006-01-28 04:54:17Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Class::DBI::Plugin::DateTime::Pg;
use strict;
use base qw(Class::DBI::Plugin::DateTime::Base);
use DateTime::Format::Pg;

BEGIN
{
    # Look ma, I can auto-generate all these :)
    my @types = qw(datetime timestamp timestamptz time timetz date duration);
    foreach my $type (@types) {
        my @args = ($type) x 3;
        push @args, $type eq 'duration' ? ", 'DateTime::Duration'" : '';
        eval sprintf(<<'        EOM', @args);
            sub has_%s
            {
                my $class  = shift;
                my $column = shift;
                my $opts   = shift || {};

                my @args   = exists $opts->{constructor_args} ?
                    @{$opts->{constructor_args}} : ();

                my $fmt     = DateTime::Format::Pg->new(@args);
                my $inflate = sub { $fmt->parse_%s(shift) };
                my $deflate = sub { $fmt->format_%s(shift) };
                __PACKAGE__->_setup_column($class, $column, $inflate, $deflate%s);
            }
        EOM
    }

    {
        no strict 'refs';
        *has_interval = \&has_duration;

        my @methods = map { ("has_$_") } @types;
        *_export_methods = sub { @methods };
    }
}

1;

__END__

=head1 NAME

Class::DBI::Plugin::DateTime::Pg - Use DateTime With PostgreSQL

=head1 SYNOPSIS

  package MyCDBI;
  use base qw(Class::DBI);
  use Class::DBI::Plugin::DateTime::Pg;

  __PACKAGE__->set_db(...);
  __PACKAGE__->table(...);
  __PACKAGE__->has_timestamp('a_timestamp');
  __PACKAGE__->has_date('a_date');
  __PACKAGE__->has_time('a_time');

=head1 DESCRIPTION

Class::DBI::Plugin::DateTime::Pg provides methods to work with DateTime
objects in a Class::DBI + PostgreSQL environment.

=head1 METHODS

All methods take the target column name. You may optionally specify a hashref
as the second argument. For this module, you may specify the following:

=over 4

=item constructor_args

An arrayref of arguments to be passed to the DateTime::Format::Pg object
that is used to parse/format the field.

=back

=head2 has_timestamp

=head2 has_datetime

=head2 has_date

=head2 has_time

=head2 has_timestamptz

=head2 has_timetz

=head2 has_interval

=head2 has_duration

=head1 SEE ALSO

L<DateTime|DateTime> L<DateTime::Format::Pg|DateTime::Format::Pg> L<Class::DBI>

=head1 AUTHOR

Copyright (c) 2005 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>. All rights reserved.

Development funded by Brazil Ltd E<lt>http://b.razil.jpE<gt>

=cut


lib/Class/DBI/Plugin/DateTime/Base.pm
# $Id: Base.pm 9 2006-01-28 04:54:17Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Class::DBI::Plugin::DateTime::Base;
use strict;
use DateTime;

sub import
{
    my $class = shift;
    $class->SUPER::import(@_);

    my($caller) = caller();
    $class->_do_export($caller);
}

sub _export_methods { return () }
sub _do_export
{
    my $class   = shift;
    my $caller  = shift;
    my @methods = $class->_export_methods();
    foreach my $method (@methods) {
        no strict 'refs';
        *{"${caller}::${method}"} = *{"${class}::${method}"};
    }
}

sub _setup_column
{
    my $class   = shift;
    my $target  = shift;
    my $column  = shift;
    my $inflate = shift;
    my $deflate = shift;
    my $coltype = shift || 'DateTime';

    if (! $target->can('has_lazy')) {
        eval <<"        EOM";
            package $target;
            use Class::DBI::LazyInflate;
        EOM
        die if $@;
    }

    $target->has_lazy(
        $column => $coltype,
        inflate => $inflate,
        deflate => $deflate,
    );
}

1;

__END__

=head1 NAME 

Class::DBI::Plugin::DateTime::Base - Base Class For DateTime Plugin

=head1 SYNOPSIS

   package MyPlugin;
   use base qw(Class::DBI::Plugin::DateTime::Base);

=head1 DESCRIPTION

Base class for Class::DBI::Plugin::DateTime classes.

=head1 AUTHOR

Copyright (c) 2005 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>. All rights reserved.

Development funded by Brazil Ltd E<lt>http://b.razil.jpE<gt>

=cut



lib/Class/DBI/Plugin/DateTime/MySQL.pm
# $Id: MySQL.pm 4 2005-11-17 06:33:36Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Class::DBI::Plugin::DateTime::MySQL;
use strict;
use base qw(Class::DBI::Plugin::DateTime::Base);
use DateTime::Format::MySQL;

BEGIN
{
    my @types = qw(datetime date timestamp);
    foreach my $type (@types) {
        # DT::F::MySQL doesn't have a format_timestamp method, so change
        # the deflator method accordingly
        my($inflator, $deflator);
        if ($type eq 'timestamp') {
            $inflator = 'timestamp';
            $deflator = 'datetime';
        } else {
            $inflator = $type;
            $deflator = $type;
        }

        eval sprintf(<<'        EOM', $type, $inflator, $deflator);
            sub has_%s
            {
                my $class  = shift;
                my $column = shift;

                my $inflate = sub { DateTime::Format::MySQL->parse_%s(shift) };
                my $deflate = sub { DateTime::Format::MySQL->format_%s(shift) };
                __PACKAGE__->_setup_column($class, $column, $inflate, $deflate);
            }
        EOM
    }

    {
        my @methods = map { ("has_$_") } @types;
        *_export_methods = sub { @methods };
    }
}

1;

__END__

=head1 NAME

Class::DBI::Plugin::DateTime::MySQL - Use DateTime With MySQL

=head1 SYNOPSIS

  package MyCDBI;
  use base qw(Class::DBI);
  use Class::DBI::Plugin::DateTime::MySQL;

  __PACKAGE__->set_db(...);
  __PACKAGE__->table(...);
  __PACKAGE__->has_timestamp('a_timestamp');
  __PACKAGE__->has_date('a_date');
  __PACKAGE__->has_time('a_time');

=head1 DESCRIPTION

Class::DBI::Plugin::DateTime::MySQL provides methods to work with DateTime
objects in a Class::DBI + MySQL environment.

=head1 METHODS

=head2 has_timestamp

=head2 has_datetime

=head2 has_date

=head1 SEE ALSO

L<DateTime|DateTime> L<DateTime::Format::MySQL|DateTime::Format::MySQL> L<Class::DBI>

=head1 AUTHOR

Copyright (c) 2005 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>. All rights reserved.

Development funded by Brazil Ltd E<lt>http://b.razil.jpE<gt>

=cut



Build.PL
# $Id: Build.PL 9 2006-01-28 04:54:17Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

use strict;
use Module::Build;

print <<EOM;

We want to know which of the bundled modules you intend to use.
All modules will be installed anyway, but if you answer "no" to any of the
questions below, that particular database type will not be tested, nor
will the dependencies be required.

Do you intend to use ...
EOM

my @exclude_tests;
my %requires = (
    'Class::DBI' => 0,
    'Class::DBI::LazyInflate' => 0,
    'Module::Pluggable' => 0
);
my %depends = (
    pg    => {
        module   => 'Class::DBI::Plugin::DateTime::Pg',
        requires => { 'DBD::Pg' => 0, 'DateTime::Format::Pg' => 0 },
        env      => [
            POSTGRES_DSN      => 'PostgreSQL DSN to use for testing (empty to skip test)?',
            POSTGRES_USER     => 'PostgreSQL username to use for testing?',
            POSTGRES_PASSWORD => 'PostgreSQL password to use for testing?',
        ],
    },
    mysql => {
        module   => 'Class::DBI::Plugin::DateTime::MySQL',
        requires => { 'DBD::mysql' => 0, 'DateTime::Format::MySQL' => 0 },
        env      => [
            MYSQL_DSN      => 'MySQL DSN to use for testing (empty to skip test)?',
            MYSQL_USER     => 'MySQL username to use for testing?',
            MYSQL_PASSWORD => 'MySQL password to use for testing?',
        ],
    },
);

my %dsn;
while (my($type, $data) = each %depends) {
    my $skip = 1;

    if (Module::Build->y_n($data->{module} . '?', "y")) {
        $skip = 0;
        my $deps = $data->{requires};
        while (my($module, $version) = each %$deps) {
            $requires{$module} = $version;
        }
    }

DONE:
    if ($skip) {
        push @exclude_tests, '\b' . $type . '.*\\.t\z';
    }
}

my $code = <<'EOM';
sub ACTION_disttest
{
    my $self = shift;
    local $ENV{PERL_TESTPOD} = 1;
    $self->SUPER::ACTION_disttest(@_);
}

sub ACTION_test
{
    my $self = shift;
EOM
while (my($key, $value) = each %dsn) {
    $code .= sprintf(qq|    local \$ENV{%s} = '%s';\n|, $key, $value);
}
$code .= <<'EOM';
    $self->SUPER::ACTION_test(@_);
}

sub find_test_files
{
    my $self  = shift;
    my $tests = $self->SUPER::find_test_files(@_);
EOM

if (scalar @exclude_tests) {
    $code .= sprintf(<<'EOM', join('|', @exclude_tests));
    $tests = [ grep { !/%s/ } @$tests ];
EOM
}

$code .= <<'EOM';
    return $tests if $ENV{PERL_TESTPOD};
    return [ grep { $_ !~ /\bpod.*\.t\z/ } @$tests ];
}
EOM

my $class = Module::Build->subclass(code => $code);

my $build = $class->new(
    dist_name     => 'Class-DBI-Plugin-DateTime',
    dist_author   => 'Daisuke Maki <dmaki@cpan.org>',
    dist_version_from => 'lib/Class/DBI/Plugin/DateTime.pm',
    dist_abstract => 'Use DateTime Objects As Your Class::DBI Columns',
    license       => 'perl',
    requires      => \%requires,
    create_makefile_pl => 'traditional',
);

$build->create_build_script();